[REVIEW] Naive Bayes

cjnolet opened the pull request:
Originally, this PR was PyTorch-based and was going to be placed in an experimental package. I have since found cuPy's API for sparse by sparse matrix multiplication and have moved the implementation our of the experimental package. 

This implementation is now 100% cupy-based. One of the downsides to having supported a PyTorch-based implementation is a known memory leak in DLPack.

cjnolet commented:
Remaining tasks:
- [ ] Double precision pytests

cjnolet commented:
rerun tests

cjnolet commented:
rerun tests

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+            The columns correspond to the classes in sorted order, as they
+            appear in the attribute classes_.
+        """
+        return cp.exp(self.predict_log_proba(X))
```
rmm_cupy_ary

cjnolet replied:
I'll have to check this, but I believe this is returning a float on host.

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+        self.class_count_ = cp.zeros(n_effective_classes, order="F",
+                                     dtype=dtype)
+        self.feature_count_ = cp.zeros((n_effective_classes, n_features),
+                                       order="F", dtype=dtype)
```
Use rmm_cupy_ary

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+        counts = cp.zeros((self.n_classes_, self.n_features_),
+                          order="F", dtype=X.dtype)
+
+        class_c = cp.zeros(self.n_classes_, order="F", dtype=X.dtype)
```
Use rmm_cupy_ary

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+                raise ValueError("Number of classes must match "
+                                 "number of priors")
+
+            self.class_log_prior_ = cp.log(class_prior)
```
rmm_cupy_ary

cjnolet replied:
Same as above, I believe this should be inheriting the allocator from `class_prior`

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+        elif self.fit_prior:
+            log_class_count = cp.log(self.class_count_)
+            self.class_log_prior_ = log_class_count - \
+                cp.log(self.class_count_.sum())
```
rmm_cupy_ary

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+                cp.log(self.class_count_.sum())
+        else:
+            self.class_log_prior_ = cp.full(self.n_classes_,
+                                            -math.log(self.n_classes_))
```
rmm_cupy_ary

dantegd commented on python/cuml/preprocessing/label.py:
```diff
@@ -0,0 +1,125 @@
+    if not check_labels(labels, classes):
+        raise ValueError("Unseen classes encountered in input")
+
+    row_ind = cp.arange(0, labels.shape[0], 1, dtype=y.dtype)
```
rmm_cupy_ary

dantegd commented on python/cuml/preprocessing/label.py:
```diff
@@ -0,0 +1,125 @@
+            y_mapped = y.tocsr().indices.astype(self.classes_.dtype)
+        else:
+            y_mapped = cp.argmax(
+                cp.asarray(y, dtype=y.dtype), axis=1).astype(y.dtype)
```
rmm_cupy_ary

dantegd commented on python/cuml/prims/__init__.py:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
```
```suggestion
# Copyright (c) 2020, NVIDIA CORPORATION.
```

dantegd commented on python/cuml/prims/label/__init__.py:
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
```
```suggestion
# Copyright (c) 2020, NVIDIA CORPORATION.
```

dantegd commented on python/cuml/prims/label/classlabels.py:
```diff
@@ -0,0 +1,236 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
```
```suggestion
# Copyright (c) 2020, NVIDIA CORPORATION.
```

dantegd commented on python/cuml/test/dask/test_label_binarizer.py:
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
```
```suggestion
# Copyright (c) 2020, NVIDIA CORPORATION.
```

raydouglass commented on CHANGELOG.md:
```diff
@@ -84,6 +85,7 @@
 - PR #1307: Add RF regression benchmark
 - PR #1327: Update the code to build treelite with protobuf
 - PR #1289: Add Python benchmarking support for FIL
+- PR #1370: Add LabelBinarizer and MNMG LabelBinarizer
```
I think this should be removed right?

cjnolet replied:
Yeah, I think we can remove it. I kept it originally to maintain a reference back to the original PR, but then the code was updated in this PR.

cjnolet commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,205 @@
+    with discrete features (e.g., word counts for text classification).
+
+    The multinomial distribution normally requires integer feature counts.
+    However, in practice, fractional counts such as tf-idf may also work.
```
I'd like to add an example in this PR if I can, though I just realized one of the reasons I haven't added the example yet is because of the hacky way I needed to create the sparse cupy-backed dask.array from the scipy sparse array using the 20newgroups dataset. 

We have a few options (and possibly more):
1. Not add an example until we can reliably create a sparse cupy-backed Dask.Array from a cupy.sparse array.
2. Add a utility for now that converts a cpu-based scipy.sparse array to a cupy-backed sparse Dask.Array. 
3. Wait until Dask and CuPY properly support the needed scipy.sparse features to use Dask.array. 

#3 could take awhile so I think that might be out of the question. I hesitate to do #1 as I'm concerned it might mean doing #3. I could do #2 in the meantime. It's unfortunate, but it might be the most reasonable solution for now, given the first-class support is still pending. What do you think?

dantegd replied:
I think you're correct, solution 2 seems to offer the best balance and it might be pretty useful while 3 happens entirely. It also "shields" us from bugs that could be introduced in the new cupy sparse features. Do you have a feel of what could be the expected timeline of the features?

cjnolet commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,205 @@
+
+        n_features = X.shape[1]
+
+        classes = cp.unique(y.map_blocks(cp.unique).compute()) \
```
It would probably make sense to use it for the local ones on the workers too, right? Do you know if the serialization (and UCX support) supporting RMM allocated memory yet?

cjnolet commented on python/cuml/dask/preprocessing/label.py:
```diff
@@ -0,0 +1,120 @@
+
+
+def cp_to_df(cp_ndarr, sparse):
+    numba_arr = numba.cuda.to_device(cp_ndarr)
```
You are correct, I'm just noticing I misread the Numba docs for this function.

cjnolet commented:
The unfortunate thing is that there are several moving parts that depend on
other projects. We need to get the CuPy vstack/hstack PR merged, then fix
the slicing issues in CSR so that the proper tests can be written in Dask
and that PR merged.

The CuPy team also seems really busy. It's taken over a month to get a
second round of review on the stack PR. The work the slicing hasn't even
began yet. It might be a little while, unfortunately. In the meantime, we
can supply utility functions.

On Thu, Feb 6, 2020 at 9:17 PM Dante Gama Dessavre <notifications@github.com>
wrote:

> *@dantegd* commented on this pull request.
> ------------------------------
>
> In python/cuml/dask/naive_bayes/naive_bayes.py
> <https://github.com/rapidsai/cuml/pull/1375#discussion_r376184104>:
>
> > +from cuml.dask.common import extract_arr_partitions, \
> +    workers_to_parts
> +
> +from dask.distributed import default_client
> +
> +
> +class MultinomialNB(object):
> +
> +    """
> +    Distributed Naive Bayes classifier for multinomial models
> +
> +    The multinomial Naive Bayes classifier is suitable for classification
> +    with discrete features (e.g., word counts for text classification).
> +
> +    The multinomial distribution normally requires integer feature counts.
> +    However, in practice, fractional counts such as tf-idf may also work.
>
> I think you're correct, solution 2 seems to offer the best balance and it
> might be pretty useful while 3 happens entirely. It also "shields" us from
> bugs that could be introduced in the new cupy sparse features. Do you have
> a feel of what could be the expected timeline of the features?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rapidsai/cuml/pull/1375?email_source=notifications&email_token=AAJPKYAMRUUZ4LPE4PLQ2EDRBTADLA5CNFSM4JOAY2CKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCUTUTSY#discussion_r376184104>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJPKYC2RBU6F3MD3TXKA6LRBTADLANCNFSM4JOAY2CA>
> .
>

cjnolet commented:
Experiencing some strange segfaults that appear to be very similar to some other behaviors we've encountered in other projects recently. Filed an issue with the CuPy team to gather their thoughts:

cjnolet commented:
@dantegd, I believe this *should* be ready for final review. Took a lot of debugging to get to the bottom of (and fix) the crash issue w/ sparse cupy dask.array but I think this is finally done.

cjnolet commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+
+        logsumexp = cp.log(cp.sum(cp.exp(jll - a_max), axis=1))
+
+        a_max = cp.squeeze(a_max, axis=1)
```
Interesting. Do these not inherit the allocator from the ndarrays they are wrapping?

cjnolet replied:
Indeed, it looks like the allocator is not being inherited. What are your thoughts on having one of our `__init__.py` files call `cp.cuda.set_allocator(rmm.rmm_cupy_allocator)`? This might alleviate the need to have to do it manually everywhere.

cjnolet commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,533 @@
+        """
+
+        jll = self._joint_log_likelihood(X)
+        indices = cp.argmax(jll, axis=1).astype(self.classes_.dtype)
```
I might be misunderstanding the use of `rmm_cupy_ary` but I thought it only needed to be used for new buffer allocations in arrays that weren't just wrapping another array. I've been assuming CuPy in inheriting the allocator from wrapped classes, otherwise it would be strange if it's running kernels on memory that's allocated in two different pools. We might need to check what it's doing under the hood.

dantegd commented on python/cuml/dask/common/__init__.py:
```diff
@@ -21,8 +21,28 @@
+import copyreg
+
+
+def serialize_mat_descriptor(m):
```
should we have a todo or comment here regarding the current need for this and that it's temporary?

cjnolet replied:
That’s a good idea. I’ll link to the cupy issue.

dantegd commented on python/cuml/dask/common/dask_arr_utils.py:
```diff
@@ -0,0 +1,180 @@
+
+    Parameters
+    ----------
+    cudf_or_array : cuDF, array-like sparse / dense array, or
```
```suggestion
    cudf_or_array : cuDF DataFrame, array-like sparse / dense array, or
```

dantegd commented on python/cuml/dask/common/dask_arr_utils.py:
```diff
@@ -0,0 +1,180 @@
+    if isinstance(cudf_or_array, dask.dataframe.DataFrame):
+        """
+        Dask.Dataframe needs special attention since it has multiple dtypes.
+        Just use the first (and assume all the rest are the same)
```
Could we have a gh issue to look into having a check that they are indeed the same?

cjnolet replied:
We could do a unique here just to verify. That Would work

dantegd commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,297 @@
+#
+
+import cupy as cp
+
```
I'm surprised flake/pep didn't whine about this two blank lines instead of one

dantegd commented on python/cuml/dask/preprocessing/label.py:
```diff
@@ -0,0 +1,233 @@
+    @staticmethod
+    def _func_inv_xform(model, y, threshold):
+        y = cp.asarray(y, dtype=y.dtype)
+        return model.inverse_transform(y, threshold)
```
should this 3 be using rmm_cupy_ary?

cjnolet replied:
Yep, missed these.

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,563 @@
+
+        self.alpha = alpha
+        self.fit_prior = fit_prior
+        self.class_prior = class_prior
```
I think class prior might benefit from using `input_to_dev_array`

dantegd commented on python/cuml/naive_bayes/naive_bayes.py:
```diff
@@ -0,0 +1,563 @@
+            inds = rmm_cupy_ary(cp.asarray, X.indices, dtype=X.indices.dtype)
+            indptr = rmm_cupy_ary(cp.asarray, X.indptr, dtype=X.indptr.dtype)
+            data = rmm_cupy_ary(cp.asarray, X.data, dtype=X.data.dtype)
+            X = cp.sparse.csr_matrix((data, inds, indptr), shape=X.shape)
```
This could be a future feature/follow up improvement, but similar to my comment above: Why _partial_fit only accepts host data?

dantegd replied:
discussion offline: future work

cjnolet replied:
This was actually a really simple fix so I just went ahead and did it.

dantegd commented on python/cuml/test/dask/test_nearest_neighbors.py:
```diff
@@ -88,7 +88,7 @@ def test_compare_skl(nrows, ncols, nclusters, n_parts, n_neighbors,
         wait(X_cudf)
 
-        cumlModel = daskNN(verbose=False, n_neighbors=n_neighbors,
+        cumlModel = daskNN(verbose=True, n_neighbors=n_neighbors,
```
why this change in this PR? I think it might've slipped in from debugging the dask nn issue?

cjnolet commented:
@dantegd, I've addressed all your review feedback, if you have a moment for a quick re-review. I'm hoping we might be able to get this in tonight, if possible.

cjnolet commented:
rerun tests

cjnolet commented:
rerun tests

cjnolet commented:
rerun test

cjnolet commented:
rerun tests

cjnolet commented:
rerun tests
